# Exercise 4.8

# STEP 1. Importing libraries and dataframe

In [1]:
# Importing libraries
import numpy as np
import pandas as pd
import os

In [2]:
# Setting the path for the datasets
path = r'C:\\Users\\cpaul\\Documents\\Career Foundry\\Instacart Basket Analysis'

#STEP 3: Import dataframe

# Importing orders_products_merged_updated dataframe
df_orders_products_merged_updated = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))


In [3]:
# Check the dataframe df_orders_products_merged_updated
df_orders_products_merged_updated.shape

(32404859, 18)

In [4]:
# Checking the columns
df_orders_products_merged_updated.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_previous_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', '_merge', 'price_range_loc', 'busiest_day',
       'busiest_days', 'busiest_period_of_day'],
      dtype='object')

# STEP 2. Calculate the aggregated mean of the order_number column grouped by department_id

In [6]:
df_orders_products_merged_updated.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# STEP 3: Analyze the results for the entire dataframe. 

# The aggregated mean is listed for all the departments when the entire dataframe is used whereas only some departments were included when the subset was used. The mean values represent all the orders for every department, so there is less variability in the mean values for the entire dataframe as opposed to the subset. The mean values were totally skewed for each department in the subset since all the orders were not included. 

# STEP 4: Creating loyalty flags for existing customers

In [8]:
# Create the max_order column which stores the the maximum order number for each user. 
# So, the dataframe is first grouped by user_id and then the transform() function is applied on the order_number column with np.max argument

df_orders_products_merged_updated['max_order'] = df_orders_products_merged_updated.groupby(['user_id'])['order_number'].transform(np.max)

df_orders_products_merged_updated.head(20)


order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_previous_order  product_id  add_to_cart_order  reordered  \
0                         NaN         196                  1          0   
1                        15.0         196                  1          1   
2                        21.0         196                  1          1   
3                        29.0         196                  1          1   
4                        28.0         196                  1          1   
5                        19.0         196                  1          1   
6                        20.0         196                  1          1   
7                        14.0         196                  2          1   
8                         0.0         196                  4          1   
9                        30.0         196                  1          1   
10                        7.0         196                  2          0   
11                        8.0         196                  6          1   
12                        7.0         196                  1          1   
13                        7.0         196                  1          1   
14                       14.0         196                  1          1   
15                        6.0         196                  3          0   
16                        8.0         196                 15          1   
17                        6.0         196                  1          1   
18                        6.0         196                  2          0   
19                        7.0         196                 17          0   

   product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0          Soda        77              7     9.0   both  Mid-range product   
1          Soda        77              7     9.0   both  Mid-range product   
2          Soda        77              7     9.0   both  Mid-range product   
3          Soda        77              7     9.0   both  Mid-range product   
4          Soda        77              7     9.0   both  Mid-range product   
5          Soda        77              7     9.0   both  Mid-range product   
6          Soda        77              7     9.0   both  Mid-range product   
7          Soda        77              7     9.0   both  Mid-range product   
8          Soda        77              7     9.0   both  Mid-range product   
9        

In [10]:
# Create a loyalty flag using the loc() function

df_orders_products_merged_updated.loc[df_orders_products_merged_updated['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
df_orders_products_merged_updated.loc[(df_orders_products_merged_updated['max_order'] <= 40) & (df_orders_products_merged_updated['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
df_orders_products_merged_updated.loc[(df_orders_products_merged_updated['max_order'] <= 10), 'loyalty_flag'] = 'New customer'

# Check the frequency of the loyalty_flag column

df_orders_products_merged_updated['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [11]:
df_orders_products_merged_updated[['user_id','loyalty_flag','order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

# STEP 5: Determine whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [12]:
# Check the basic statistics of the product prices for each loyalty category
df_orders_products_merged_updated.groupby('loyalty_flag').agg({'prices': ['mean','min','max','median']})

prices                     
                       mean  min      max median
loyalty_flag                                    
Loyal customer    10.386336  1.0  99999.0    7.4
New customer      13.294670  1.0  99999.0    7.4
Regular customer  12.495717  1.0  99999.0    7.4

# The mean of the product price for Loyal customers seem to be lower than the regular customers, also there are more regular customers than loyal customers. The number of New customers are lesser than the other two groups, so we cannot fully take into account the mean of product price of this group even though the mean is the highest.

# STEP 6: Target different types of spenders in their marketing campaigns

In [14]:
# Create a spending flag for each user based on the average price across all their orders

df_orders_products_merged_updated['mean_price'] = df_orders_products_merged_updated.groupby(['user_id'])['prices'].transform(np.mean)

df_orders_products_merged_updated.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_previous_order  product_id  add_to_cart_order  reordered  \
0                         NaN         196                  1          0   
1                        15.0         196                  1          1   
2                        21.0         196                  1          1   
3                        29.0         196                  1          1   
4                        28.0         196                  1          1   
5                        19.0         196                  1          1   
6                        20.0         196                  1          1   
7                        14.0         196                  2          1   
8                         0.0         196                  4          1   
9                        30.0         196                  1          1   
10                        7.0         196                  2          0   
11                        8.0         196                  6          1   
12                        7.0         196                  1          1   
13                        7.0         196                  1          1   
14                       14.0         196                  1          1   
15                        6.0         196                  3          0   
16                        8.0         196                 15          1   
17                        6.0         196                  1          1   
18                        6.0         196                  2          0   
19                        7.0         196                 17          0   

   product_name  ...  department_id  prices  _merge    price_range_loc  \
0          Soda  ...              7     9.0    both  Mid-range product   
1          Soda  ...              7     9.0    both  Mid-range product   
2          Soda  ...              7     9.0    both  Mid-range product   
3          Soda  ...              7     9.0    both  Mid-range product   
4          Soda  ...              7     9.0    both  Mid-range product   
5          Soda  ...              7     9.0    both  Mid-range product   
6          Soda  ...              7     9.0    both  Mid-range product   
7          Soda  ...              7     9.0    both  Mid-range product   
8          Soda  ...              7     9.0    both  Mid-range product   
9          Soda  ...              7     9.0    bo

In [15]:
# Create a spender_type flag using the loc() function

df_orders_products_merged_updated.loc[df_orders_products_merged_updated['mean_price'] >= 10, 'spender_type_flag'] = 'High spender'
df_orders_products_merged_updated.loc[(df_orders_products_merged_updated['mean_price'] < 10), 'spender_type_flag'] = 'Low spender'

# Check the frequency of the loyalty_flag column

df_orders_products_merged_updated['spender_type_flag'].value_counts(dropna = False)

Low spender     31770614
High spender      634245
Name: spender_type_flag, dtype: int64

# STEP 7: Determine frequent versus non-frequent customers  to send relevant notifications within the app

In [20]:
# Create a frequency flag for each user based on the median of days_since_previous_order across all their orders

df_orders_products_merged_updated['median_days_since_previous_order'] = df_orders_products_merged_updated.groupby(['user_id'])['days_since_previous_order'].transform(np.median)

df_orders_products_merged_updated.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_previous_order  product_id  add_to_cart_order  reordered  \
0                         NaN         196                  1          0   
1                        15.0         196                  1          1   
2                        21.0         196                  1          1   
3                        29.0         196                  1          1   
4                        28.0         196                  1          1   
5                        19.0         196                  1          1   
6                        20.0         196                  1          1   
7                        14.0         196                  2          1   
8                         0.0         196                  4          1   
9                        30.0         196                  1          1   
10                        7.0         196                  2          0   
11                        8.0         196                  6          1   
12                        7.0         196                  1          1   
13                        7.0         196                  1          1   
14                       14.0         196                  1          1   
15                        6.0         196                  3          0   
16                        8.0         196                 15          1   
17                        6.0         196                  1          1   
18                        6.0         196                  2          0   
19                        7.0         196                 17          0   

   product_name  ...  _merge    price_range_loc     busiest_day  \
0          Soda  ...    both  Mid-range product  Regularly busy   
1          Soda  ...    both  Mid-range product  Regularly busy   
2          Soda  ...    both  Mid-range product  Regularly busy   
3          Soda  ...    both  Mid-range product      Least busy   
4          Soda  ...    both  Mid-range product      Least busy   
5          Soda  ...    both  Mid-range product  Regularly busy   
6          Soda  ...    both  Mid-range product  Regularly busy   
7          Soda  ...    both  Mid-range product  Regularly busy   
8          Soda  ...    both  Mid-range product  Regularly busy   
9          Soda  ...    both  Mid-range product      Least busy   
10         Soda  ...    both  Mid-range product  Reg

In [21]:
# Create a frequency flag using the loc() function

df_orders_products_merged_updated.loc[df_orders_products_merged_updated['median_days_since_previous_order'] > 20, 'frequency_flag'] = 'Non-frequent customer'
df_orders_products_merged_updated.loc[(df_orders_products_merged_updated['median_days_since_previous_order'] <= 20) & (df_orders_products_merged_updated['median_days_since_previous_order'] > 10), 'frequency_flag'] = 'Regular customer'
df_orders_products_merged_updated.loc[(df_orders_products_merged_updated['median_days_since_previous_order'] <= 10), 'frequency_flag'] = 'Frequent customer'

# Check the frequency of the frequency_flag column

df_orders_products_merged_updated['frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: frequency_flag, dtype: int64

In [22]:
# Checking the Nan values
print(df_orders_products_merged_updated[df_orders_products_merged_updated['frequency_flag'].isnull()])

          order_id  user_id  order_number  orders_day_of_week  \
13645692    895835   159838             1                   0   
17251990    895835   159838             1                   0   
17622767    895835   159838             1                   0   
24138593    895835   159838             1                   0   
25880002    895835   159838             1                   0   

          order_hour_of_day  days_since_previous_order  product_id  \
13645692                 17                        NaN       10749   
17251990                 17                        NaN       33401   
17622767                 17                        NaN       23695   
24138593                 17                        NaN       21334   
25880002                 17                        NaN       22198   

          add_to_cart_order  reordered  \
13645692                  3          0   
17251990                  6          0   
17622767                  2          0   
24138593            

# This user with user_id 159838 has only one order and days_since_previous_order column is null due to this, and, the median cannot be calculated. So, I would like to flag this customer as a Non-frequent customer.

In [24]:
# Updating the frequncy_flag for this customer
df_orders_products_merged_updated['frequency_flag'].fillna("Non-frequent customer", inplace = True)

df_orders_products_merged_updated['frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636442
Name: frequency_flag, dtype: int64

# STEP 9: Export dataframe as a pickle file.

In [25]:
#Export it in pkl format as orders_products_merged_aggregated.pkl

df_orders_products_merged_updated.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_aggregated.pkl'))